Add duration property to the Item model to support the new duration feature.

AhmedZaki99 opened the pull request:
Fixes #34

Changes include:

- Adding a [`Duration`](https://github.com/AhmedZaki99/todoist-net/blob/patch-3/src/Todoist.Net/Models/Duration.cs) model to be mapped against the task duration JSON object.
- Adding a [`DurationUnit`](https://github.com/AhmedZaki99/todoist-net/blob/patch-3/src/Todoist.Net/Models/DurationUnit.cs) model that extends [`StringEnum`](https://github.com/olsh/todoist-net/blob/master/src/Todoist.Net/Models/StringEnum.cs) and supports two values: minute & day.
- Adding a [`Duration`](https://github.com/AhmedZaki99/todoist-net/blob/patch-3/src/Todoist.Net/Models/Item.cs#L116) property of type `Duration` in the [`Item`](https://github.com/olsh/todoist-net/blob/master/src/Todoist.Net/Models/Item.cs) model.

olsh commented on src/Todoist.Net/Models/Duration.cs:
```diff
@@ -0,0 +1,80 @@
+                {
+                    return TimeSpan.FromDays(Amount);
+                }
+                throw new NotImplementedException();
```
I'd suggest to change ifs to switch pattern in this way static analyzers will find an issue if we add a new DurationUnit in the future.

AhmedZaki99 replied:
You have a good point, but the problem is that we are dealing with static values that won't work with switch patterns.
I've looked it up and found a way to make the code more readable, following this [SO answer](https://stackoverflow.com/a/49624848/6668824), I did the following:
```cs
switch (Unit)
{
    case var _ when Unit == DurationUnit.Minute:
        return TimeSpan.FromMinutes(Amount);
    case var _ when Unit == DurationUnit.Day:
        return TimeSpan.FromDays(Amount);
    default:
        throw new NotImplementedException();
}
```
Unfortunately, this still doesn't guide static analyzers to help us add support for new values.

I thought of adding an `enum` that can be mapped to each of the `DurationUnit` values, but again, analyzers will be useful only if new values are added both to the class and the enum.

Maybe there's something that I'm missing, what do you think?

olsh reviewed (changes requested):
Could you please write at least one integration tests with duration?

olsh commented:
Thank you for the contribution! 👍🏻

AhmedZaki99 commented:
> Could you please write at least one integration tests with duration?

You were right about this, the test I wrote was failing at first, but then I found that the `Duration` property was being ignored when set to `null`, so I added an exception for it in the `JsonProperty` attribute:
```cs
[JsonProperty("duration", NullValueHandling = NullValueHandling.Include)]
public Duration Duration { get; set; }
```

In addition to the integration test, I wrote a couple of unit tests for the `Duration` object, which have guided me later to move the validation checks on the constructor parameters to be executed within property setters, e.g.:
```cs
public int Amount
{
    get => _amount;
    set => _amount = value > 0
        ? value
        : throw new ArgumentOutOfRangeException(nameof(Amount), "Duration amount must be greater than zero.");
}
```

All tests are passing now, but on my machine, the test I added was failing due to inconsistency in conversion between [`DueDate.StringDate`](https://github.com/olsh/todoist-net/blob/master/src/Todoist.Net/Models/DueDate.cs#L113) and [`DueDate.Date`](https://github.com/olsh/todoist-net/blob/master/src/Todoist.Net/Models/DueDate.cs#L56) properties.
When the task contains a floating due date, it is treated like a local time when being converted from `DateTime` to `string` in the `DueDate.StringDate` property getter, hence, its string value changes from what it was given by the API.

In my case (time zone is GMT +2), the value `"2021-12-22T09:15:00"` fetched from the API was converted to a `DateTime` of kind `DateTimeKind.Unspecified`, and then converted back to a `"2021-12-22T07:15:00"` when sent back to the API for update.
That results in an unwanted update on the task without even changing any field.

I think I might've faced a similar issue while I was working on my project and it's clear now why it's happening. I will wait until we get done with this pull request and then open an issue ticket for it.
